datafusion/src/scalar.rs

```diff
@@ -424,6 +424,25 @@ macro_rules! build_array_from_option {
     }};
 }
 
+macro_rules! eq_array_general {
+    ($array:expr, $index:expr, $ARRAYTYPE:ident, $VALUE:expr, $has_nulls:expr) => {{
+        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
+        if *$has_nulls {
```

Comment:
This branch should be ideally be outside of the loop of 
https://github.com/apache/arrow-datafusion/pull/922/files#diff-03876812a8bef4074e517600fdcf8e6b49f1ea24df44905d6d806836fd61b2a8L360

But this might be hard to accomplish given the current design. I posted some ideas in the earlier comment